### PR TITLE
[map] Add geotiff

### PR DESCRIPTION
**Pull request by chejennifer**
- initial version of drawing geotiff in map explorer (does NOT have feature parity with original d3 maps)
- drawing of leaflet map is only allowed with the parameter: &l=1
- some refactoring
    - took map drawing functionality out of Chart component and put it into D3Map and LeafletMap
    - broke down draw_choropleth.ts to draw_d3_map.ts and draw_map_utils.ts

screen-recording-_51_

tracking issue for tasks needed for feature parity between leaflet maps and d3 maps: https://github.com/datacommonsorg/website/issues/1774

### CODE REVIEW

**Review by beets — APPROVED** [review]
looks good - and thanks for figuring out how to incorporate leaflet. it's a little hard to follow along since it's still in flux and awaiting merging..

most comments were nits and around keeping things stable for prod maps (so we don't load extra resources unless necessary).

it would be worth having an issue with a checklist of items (or TODO's in code) that are not at feature parity yet so it's easier to review.

**Review by shifucun — APPROVED** [review]
Overall looks good. With the big refactor going on, we can check this in and fine tune the data fetch logic.

File: static/js/tools/map/chart.tsx [hunk 1]
```diff
@@ -82,6 +60,7 @@ interface ChartProps {
   europeanCountries: Array<NamedPlace>;
   rankingLink: string;
   facetInfo: FacetSelectorFacetInfo;
+  georaster: any;
```
[the remark this answers is not in the document]
Comment (chejennifer) [reply static/js/tools/map/chart.tsx]: done

File: static/js/tools/map/chart_loader.tsx [hunk 2]
```diff
@@ -181,6 +184,9 @@ export function ChartLoader(): JSX.Element {
         setRawData,
         display.value.showTimeSlider
       );
+      geoblaze
```
[the remark this answers is not in the document]
Comment (chejennifer) [reply static/js/tools/map/chart_loader.tsx]: done

File: static/js/tools/map/chart.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (chejennifer) [reply static/js/tools/map/chart.tsx]: updated the type

File: static/js/chart/draw_leaflet_map.ts [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+import * as d3 from "d3";
+import geoblaze from "geoblaze";
+import GeoRasterLayer from "georaster-layer-for-leaflet";
+import L, { geoJSON } from "leaflet";
+
+import { USA_PLACE_DCID } from "../shared/constants";
+import { getPlacePathId } from "./draw_map_utils";
+import { GeoJsonData } from "./types";
+
+const NO_DATA_VALUE = -9999;
```
[the remark this answers is not in the document]
Comment (chejennifer) [reply static/js/chart/draw_leaflet_map.ts:33]: done

File: static/js/chart/draw_leaflet_map.ts [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+const DEFAULT_MAP_ZOOM = 4;
+const GEOJSON_STYLE = { fillColor: "transparent", weight: 0.5, color: "white" };
+const GEOJSON_HIGHLIGHTED_STYLE = {
+  weight: 2,
+  color: "#666",
+  fillOpacity: 0.7,
+};
+
+function updateTooltip(
+  georaster: any,
```
[the remark this answers is not in the document]
Comment (chejennifer) [reply static/js/chart/draw_leaflet_map.ts]: updated the types

File: static/js/chart/draw_leaflet_map.ts [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+import GeoRasterLayer from "georaster-layer-for-leaflet";
+import L, { geoJSON } from "leaflet";
+
+import { USA_PLACE_DCID } from "../shared/constants";
+import { getPlacePathId } from "./draw_map_utils";
+import { GeoJsonData } from "./types";
+
+const NO_DATA_VALUE = -9999;
+const GEORASTER_DATA_BAND = 0;
+const MAP_OPTIMAL_CENTER = {
```
[the remark this answers is not in the document]
Comment (chejennifer) [reply static/js/chart/draw_leaflet_map.ts:38]: done

File: static/js/chart/draw_leaflet_map.ts [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * functions for drawing leaflet map.
```
[the remark this answers is not in the document]
Comment (chejennifer) [reply static/js/chart/draw_leaflet_map.ts]: I think high level, this file holds independent functions for each of adding a geotiff layer to a map and adding a geojson layer to a map (i.e., there's no logic saying geotiff has to be a background). I added comments in the LeafletMap component instead noting that we are drawing geotiff for the background and geojsons as the borders.

File: server/templates/tools/map.html [hunk 7]
```diff
@@ -27,7 +27,8 @@
   <link rel="stylesheet" href={{url_for('static', filename='css/map.min.css')}} >
   <link rel="stylesheet" href="https://fonts.googleapis.com/icon?family=Material+Icons">
   <link rel="stylesheet" href="https://fonts.googleapis.com/icon?family=Material+Icons+Outlined">
-{% endblock %}
+  <link rel="stylesheet" href="https://npmcdn.com/leaflet@1.0.0-rc.2/dist/leaflet.css" />
```
[the remark this answers is not in the document]
Comment (chejennifer) [reply server/templates/tools/map.html]: added the integrity checksum.

For the request arg, we use "#" to specify args for the map tool in the url & flask.request can not pick up the "#"..

[the remark this answers is not in the document]
Comment (chejennifer) [reply server/templates/tools/map.html]: so like "/tools/map/geotiff#..."?

[the remark this answers is not in the document]
Comment (beets) [reply server/templates/tools/map.html]: More like `/tools/map?l=1#<regular map params>` (no need to create a new endpoint...)
